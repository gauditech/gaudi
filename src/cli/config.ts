import path from "path";

import { EngineConfig } from "@src/config";

// -------------------- Various configs, constants and defaults

/**
 * Time to wait before reporting resource watcher changes (in millis).
 *
 * This allows basic debouncing (eg. when creating/updating multiple files at once).
 */
export const RESOURCE_WATCH_DELAY = 500;

/** Returns path to prisma schema file */
export function getDbSchemaPath(config: EngineConfig): string {
  return `${config.gaudiFolder}/db/schema.prisma`;
}

// defaul Node options
const DEFAULT_NODE_OPTIONS = [
  // development node options - maybe we should allow disabling them in production?
  "--enable-source-maps",
  "--stack-trace-limit=30",
  "--stack-size=2048",
];

/**
 * Default node options
 *
 * Additional node options can be passed via NODE_OPTIONS env var which is included when running commands
 */
export function getDefaultNodeOptions(): string[] {
  return [...DEFAULT_NODE_OPTIONS];
}

/**
 * Paths to Gaudi scripts
 *
 * Target Gaudi scripts directly instead of via NPX because we cannot pass some node options through NPX (via --node-options")
 * See a list of allowed options here: https://nodejs.org/docs/latest-v16.x/api/cli.html#node_optionsoptions
 */
export const GAUDI_SCRIPTS = {
  ENGINE: path.join(__dirname, "../engine.js"),
  RUNTIME: path.join(__dirname, "../runtime/runtime.js"),
  POPULATOR: path.join(__dirname, "../populator/populator.js"),
};

/**
 * Add path to CLI's binary.
 *
 * We don't want to force users to be aware of our binaries which means that
 * they probably won't be available as global binaries thus we have to manually
 * target them in our own `node_modules`.
 */
export function appendBinPath(binName: string): string {
  return path.join("./node_modules/@gaudi/engine/node_modules/.bin/", binName);
}
