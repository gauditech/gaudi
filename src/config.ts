import dotenv from "dotenv";
import { ConnectionOptions } from "pg-connection-string";

import { parseConnectionString } from "./common/utils";

import { GAUDI_FOLDER_NAME } from "@src/const";

export type EngineConfig = {
  /** Path to Gaudi blueprint file */
  inputPath: string;
  /** Folder where runtime should output generated files */
  outputFolder: string;
  /** Gaudi folder */
  gaudiFolder: string;
  dbConn: ConnectionOptions;
  embeddedPg: boolean;
};

/** Read runtime config from environment or provide default values. */
export function readConfig(configPath?: string): EngineConfig {
  dotenv.config({ path: configPath });

  const inputPath = process.env.GAUDI_ENGINE_INPUT_PATH ?? "";
  const outputFolder = process.env.GAUDI_ENGINE_OUTPUT_PATH ?? ".";
  // gaudi folder's path should probably be determined by the position of (future) gaudi config file
  const gaudiFolder = `./${GAUDI_FOLDER_NAME}`;

  const dbConn = parseConnectionString(process.env.GAUDI_DATABASE_URL);
  const embeddedPg = process.env.GAUDI_EMBEDDED_POSTGRESQL_ENABLED?.toLowerCase() === "true";

  return { inputPath, outputFolder, gaudiFolder, dbConn, embeddedPg };
}
