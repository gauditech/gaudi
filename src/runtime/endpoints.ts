import express, { Request } from "express";
import knex from "knex";

const app = express();
const port = 3002;

import {
  PathParam,
  buildEndpointPath,
  queryableFromEndpointTargets,
  selectToSelectable,
} from "@src/builder/query";
import { Definition, QueryDef, QueryDefPath, TargetDef } from "@src/types/definition";

function validatePathParam(param: PathParam["params"][number], val: string): PathVariable {
  switch (param.type) {
    case "integer":
      return { name: param.name, type: param.type, value: parseInt(val, 10) };
    case "text":
      return { name: param.name, type: param.type, value: val };
  }
}

type PathVariable =
  | { name: string; type: "text"; value: string }
  | { name: string; type: "integer"; value: number };

export function setupEndpoints(def: Definition) {
  for (const entrypoint of def.entrypoints) {
    for (const endpoint of entrypoint.endpoints) {
      switch (endpoint.kind) {
        case "get": {
          const { path, params } = buildEndpointPath(endpoint);
          app.get(path, (req: Request, res) => {
            // collect path params
            const vars = params.map((p) => validatePathParam(p, req.params[p.name]));

            // get targets
            const q = queryableFromEndpointTargets(
              def,
              endpoint.targets,
              selectToSelectable(endpoint.response),
              "single"
            );
          });
        }
      }
    }
  }
}

/**
 * Here we go again... Returning a query/able, whatever it may be.


from Org.public_repos.issues
filter
      Org.slug is ...
  and public_repos.slug is ...
  and     
*/
