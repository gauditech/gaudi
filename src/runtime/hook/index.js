/** Returns given value. Can be used for testing. */
module.exports.echo = function ({ value }) {
  return value;
};
