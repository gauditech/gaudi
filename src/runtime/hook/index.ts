/** Concatente all given string parts */
export function concat(...parts: string[]): string {
  return parts.join("");
}
