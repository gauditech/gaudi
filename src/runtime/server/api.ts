import path from "path";

import { Express, NextFunction, Request, Response, static as staticHandler } from "express";
import { OpenAPIV3 } from "openapi-types";
import { serve, setup } from "swagger-ui-express";

import { buildOpenAPI } from "@src/builder/openAPI";
import { saveOutputFile } from "@src/common/utils";
import { buildEntrypoints } from "@src/runtime/server/admin";
import { buildEndpoints as buildAuthEndpoints } from "@src/runtime/server/authentication";
import { getAppContext } from "@src/runtime/server/context";
import { buildEndpointConfig, registerServerEndpoint } from "@src/runtime/server/endpoints";
import { EndpointConfig } from "@src/runtime/server/types";
import { Definition, EntrypointDef } from "@src/types/definition";

/** Create endpoint handlers, OpenAPI specs and attach them to server instance */
export function setupServerApis(definition: Definition, app: Express) {
  const config = getAppContext(app).config;

  // --- static folder (eg. for API specs)
  const specOutputFolder = path.join(config.outputFolder, "api-spec");
  app.use("/api-spec", staticHandler(specOutputFolder));

  // --- definition API (aka the "blueprint")
  const definitionEntrypoints = definition.entrypoints;
  const definitionEndpointConfigs = buildEndpointConfig(definition, definitionEntrypoints);
  setupEntrypointApi(
    "api",
    definition,
    definitionEntrypoints,
    definitionEndpointConfigs,
    app,
    specOutputFolder
  );

  // --- admin API
  const adminEntrypoints = buildEntrypoints(definition);
  const adminEndpointConfigs = buildEndpointConfig(definition, adminEntrypoints);
  setupEntrypointApi(
    "api-admin",
    definition,
    adminEntrypoints,
    adminEndpointConfigs,
    app,
    specOutputFolder
  );

  // --- authentication API
  const authConfigs = buildAuthEndpoints();
  setupConfigEndpoints("api", authConfigs, app);
}

/** Create API endpoints from entrypoints.*/
function setupEntrypointApi(
  name: string,
  definition: Definition,
  entrypoints: EntrypointDef[],
  endpointConfigs: EndpointConfig[],
  app: Express,
  outputFolder: string
) {
  const basePath = `/${name}`;
  const specFile = path.join(outputFolder, `${name}.openapi.json`);

  endpointConfigs.forEach((epc) => registerServerEndpoint(app, epc, basePath));

  setupEntrypointApiSpec(definition, entrypoints, app, basePath, specFile);
}

/** Create API OpenAPI spec from entrpyoints */
function setupEntrypointApiSpec(
  definition: Definition,
  entrypoints: EntrypointDef[],
  app: Express,
  basePath: string,
  outputFile: string
) {
  const openApi = buildOpenAPI(
    { models: definition.models, entrypoints, populators: [] },
    basePath
  );

  saveOutputFile(outputFile, JSON.stringify(openApi, undefined, 2));

  setupEntrypointApiSwagger(openApi, app, basePath);
}

function setupEntrypointApiSwagger(
  openApiDocument: OpenAPIV3.Document,
  app: Express,
  apiPath: string
) {
  const swaggerPath = `${apiPath}-docs`;

  app.use(swaggerPath, serve, (_req: Request, _resp: Response, _next: NextFunction) =>
    setup(openApiDocument)(_req, _resp, _next)
  );
}

/** Setup API endpoints from endpoint configs */
function setupConfigEndpoints(name: string, configs: EndpointConfig[], app: Express) {
  const basePath = `/${name}`;

  configs.forEach((epc) => {
    registerServerEndpoint(app, epc, basePath);
  });

  // NOTE: we cannot currently create OpenAPI spec from endpoint configs
}
