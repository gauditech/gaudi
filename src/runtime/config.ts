import fs from "fs";

import { initializeContext } from "@src/runtime/server/context";
import { createDbConn } from "@src/runtime/server/dbConn";
import { Definition } from "@src/types/definition";

export type RuntimeConfig = {
  /** Runtime server host name */
  host: string;
  /** Runtime server port number */
  port: number;
  /** Path to generated definition.json file. */
  definitionPath: string;
  /** Folder where runtime should output generated files */
  outputFolder: string;
  /** DB connection URL */
  dbConnUrl: string;
  /** DB default schema */
  dbSchema?: string;
};

/** Read runtime config from environment or provide default values. */
export function readConfig(): RuntimeConfig {
  const host = process.env.GAUDI_RUNTIME_SERVER_HOST ?? "127.0.0.1";
  const port =
    process.env.GAUDI_RUNTIME_SERVER_PORT != null
      ? parseInt(process.env.GAUDI_RUNTIME_SERVER_PORT, 10)
      : 3001;
  const definitionPath = process.env.GAUDI_RUNTIME_DEFINITION_PATH || "definition.json";
  const outputFolder = process.env.GAUDI_RUNTIME_OUTPUT_PATH || ".";

  const dbConnUrl = process.env.GAUDI_DATABASE_URL || "";
  const dbSchema = process.env.GAUDI_DATABASE_SCHEMA;

  return { host, port, definitionPath, outputFolder, dbConnUrl, dbSchema };
}

/** Load definition file and return it's content */
export function loadDefinition(definitionPath: string): Definition {
  // --- read input file
  if (!fs.existsSync(definitionPath)) {
    throw new Error(`Definition file not found: "${definitionPath}"`);
  }
  const definitionStr = fs.readFileSync(definitionPath).toString("utf-8");
  return JSON.parse(definitionStr);
}

export function createAppContext(config: RuntimeConfig) {
  initializeContext({
    dbConn: createDbConn(config.dbConnUrl, {
      schema: config.dbSchema,
    }),
    config: config,
  });
}
