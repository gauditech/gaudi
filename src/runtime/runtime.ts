// import this file only with relative path because this file actually configures path aliases (eg @src, ...)
import "../common/setupAliases";

import { loadDefinition, readConfig } from "@src/runtime/config";
import { importHooks } from "@src/runtime/hooks";
import { createServer } from "@src/runtime/server/server";

// read environment
const config = readConfig();

const definition = loadDefinition(config.definitionPath);

importHooks(config);

// start server
createServer(definition, config);
