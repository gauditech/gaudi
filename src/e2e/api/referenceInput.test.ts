import path from "path";

import _ from "lodash";
import request from "supertest";

import { createApiTestSetup, loadBlueprint } from "@src/e2e/api/setup";
import { readConfig } from "@src/runtime/config";

// these tests last longer than default 5s timeout so this seems to help
jest.setTimeout(10000);

describe("Reference Input", () => {
  const config = readConfig(path.join(__dirname, "api.test.env"));

  const { getServer, setup, destroy } = createApiTestSetup(
    config,
    loadBlueprint(path.join(__dirname, "referenceInput.gaudi")),
    []
  );

  describe("Element and Extra", () => {
    beforeAll(async () => {
      await setup();
    });
    afterAll(async () => {
      await destroy();
    });

    it("create with a valid reference", async () => {
      const extraData = { slug: "extra" };

      const extraPostResponse = await request(getServer()).post("/extra").send(extraData);
      expect(extraPostResponse.statusCode).toBe(200);

      const data = { name: "element", extra_slug: "extra" };

      const postResponse = await request(getServer()).post("/element").send(data);
      expect(postResponse.statusCode).toBe(200);
      expect(postResponse.body).toMatchSnapshot();
    });

    it("create with an invalid reference", async () => {
      const data = { name: "element", extra_slug: "baz" };

      const postResponse = await request(getServer()).post("/element").send(data);
      expect(postResponse.statusCode).toBe(400);
      expect(postResponse.body).toMatchSnapshot();
    });
  });
});
