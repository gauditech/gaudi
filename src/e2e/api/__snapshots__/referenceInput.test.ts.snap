// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Reference Input Element and Extra create with extra 1`] = `
{
  "extra_id": 1,
  "id": 1,
  "name": "element",
}
`;

exports[`Reference Input Element and Extra create with extra no reference 1`] = `
{
  "code": "ERROR_CODE_VALIDATION",
  "data": {
    "errors": [
      "Can't find extra_slug with value: baz",
    ],
    "inner": [
      {
        "errors": [
          "Can't find extra_slug with value: baz",
        ],
        "inner": [],
        "message": "Can't find extra_slug with value: baz",
        "name": "ValidationError",
        "params": {
          "originalValue": "baz",
          "path": "extra_slug",
          "value": "baz",
        },
        "path": "extra_slug",
        "type": "noReference",
        "value": "baz",
      },
    ],
    "value": {
      "extra_slug": "baz",
      "name": "element",
    },
  },
  "message": "Validation error",
}
`;
